libsquoosh/src/index.ts

```diff
@@ -107,26 +131,35 @@ async function encodeImage({
   };
 }
 
-// both decoding and encoding go through the worker pool
-function handleJob(params) {
-  const { operation } = params;
-  switch (operation) {
+type EncodeParams = { operation: 'encode' } & Parameters<typeof encodeImage>[0];
+type DecodeParams = { operation: 'decode' } & Parameters<typeof decodeFile>[0];
+type PreprocessParams = { operation: 'preprocess' } & Parameters<
+  typeof preprocessImage
+>[0];
+export type JobMessage = EncodeParams | DecodeParams | PreprocessParams;
```

Comment:
Instead of exporting this from here and using it in WorkerPool, would it make sense to make WorkerPool accept a generic `JobMessage` type so that we can isolate its implementation and definition from its usage?

Reply:
Yeah that would be ideal. The code as written today has a static method (useThisThreadAsWorker) which can't use generics from the class. It also has a helper function outside the class (jobPromise) which needs to use the message as well.

Reply:
I think, we can move `jobPromise` inside the class and for now, add the generic parameters to `useThisThreadAsWorker` like `static useThisThreadAsWorker<I, O>(cb: (msg: I) => O)`. Then, we can find a better way to use the generics instead of using the static method.